scipy/integrate/_tanhsinh.py

```diff
@@ -0,0 +1,755 @@
+
+    xjcs = [_pair_cache.xjc]
+    wjs = [_pair_cache.wj]
+
+    for i in range(len(_pair_cache.indices)-1, k + 1):
+        xjc, wj = _compute_pair(i, h0)
+        xjcs.append(xjc)
+        wjs.append(wj)
+        _pair_cache.indices.append(_pair_cache.indices[-1] + len(xjc))
+
+    _pair_cache.xjc = np.concatenate(xjcs)
+    _pair_cache.wj = np.concatenate(wjs)
+    _pair_cache.h0 = h0
+
+_pair_cache.xjc = np.empty(0)
+_pair_cache.wj = np.empty(0)
+_pair_cache.indices = [0]
+_pair_cache.h0 = None
+
+
+def _get_pairs(k, h0, inclusive=False, dtype=np.float64):
+    # Retrieve the specified abscissa-weight pairs from the cache
+    # If `inclusive`, return all up to and including the specified level
+    if len(_pair_cache.indices) <= k+2 or h0 != _pair_cache.h0:
+        _pair_cache(k, h0)
+
+    xjc = _pair_cache.xjc
+    wj = _pair_cache.wj
+    indices = _pair_cache.indices
+
+    start = 0 if inclusive else indices[k]
+    end = indices[k+1]
+
+    return xjc[start:end].astype(dtype), wj[start:end].astype(dtype)
+
+
+def _transform_to_limits(xjc, wj, a, b):
+    # Transform integral according to user-specified limits. This is just
+    # math that follows from the fact that the standard limits are (-1, 1).
+    # Note: If we had stored xj instead of xjc, we would have
+    # xj = alpha * xj + beta, where beta = (a + b)/2
+    alpha = (b - a) / 2
+    xj = np.concatenate((-alpha * xjc + b, alpha * xjc + a), axis=-1)
+    wj = wj*alpha  # arguments get broadcasted, so we can't use *=
+    wj = np.concatenate((wj, wj), axis=-1)
+
+    # Points at the boundaries can be generated due to finite precision
+    # arithmetic, but these function values aren't supposed to be included in
+    # the Euler-Maclaurin sum. Ideally we wouldn't evaluate the function at
+    # these points; however, we can't easily filter out points since this
+    # function is vectorized. Instead, zero the weights.
+    invalid = (xj <= a) | (xj >= b)
+    wj[invalid] = 0
+    return xj, wj
+
+
+def _euler_maclaurin_sum(fj, work):
+    # Perform the Euler-Maclaurin Sum, [1] Section 4
+
+    # The error estimate needs to know the magnitude of the last term
+    # omitted from the Euler-Maclaurin sum. This is a bit involved because
+    # it may have been computed at a previous level. I sure hope it's worth
+    # all the trouble.
+    xr0, fr0, wr0 = work.xr0, work.fr0, work.wr0
+    xl0, fl0, wl0 = work.xl0, work.fl0, work.wl0
+
+    # It is much more convenient to work with the transposes of our work
+    # variables here.
+    xj, fj, wj = work.xj.T, fj.T, work.wj.T
+    n_x, n_active = xj.shape  # number of abscissae, number of active elements
+
+    # We'll work with the left and right sides separately
+    xr, xl = xj.reshape(2, n_x // 2, n_active).copy()  # this gets modified
+    fr, fl = fj.reshape(2, n_x // 2, n_active)
+    wr, wl = wj.reshape(2, n_x // 2, n_active)
+
+    invalid_r = ~np.isfinite(fr) | (wr == 0)
+    invalid_l = ~np.isfinite(fl) | (wl == 0)
```

Comment:
This is a line that deserves discussion in real time. 

With perfect arithmetic, the weights would be nonzero, the function values would all be finite, and the product of the function value and weight would determine the contribution to the sum. Great.

In the real world, sometimes functions are not evaluated accurately near singularities. They might even overflow to infinity well before the true value exceeds `finfo(f).max`, at a point where the weight (which may be computed more accurately than the function value) is nonzero.

I've gone the route here of assuming that non-finite values _never_ need to be considered because this allows the solution of some problems for which we'd otherwise fail. It's documented that this routine can't detect divergent integrals or handle interior singularities, so this is not so wrong. However, it's possible that this could cause us to produce inaccurate results without warning. And even if we ask the user not to rely on the routine for detecting divergent integrals or interior singularies, this line (as-is) ensures that we can't detect them.

I've considered a few ways of improving this. For example,
- Evaluate the function at the endpoints during input validation. If both are finite, then infinite values within the interval are a problem, and we shouldn't filter them out like this.
- If there is an endpoint singularity, try to distinguish between interior and endpoint infinities, and squash only interior infinities, or maybe emit warnings about interior singularities.
- Squash enpoint infinities only if weight is below some threshold.

But all of these will be heuristics, so I'll wait for the reviewer to weigh in before implementing anything.

Reply:
Yeah, it works fine for use on statistical distribution.